server/aesdsocket.c

```diff
@@ -11,144 +21,245 @@
+typedef struct thread_node {
+    pthread_t tid;
+    struct thread_node *next;
+} node_t;
 
+node_t *head = NULL;
+pthread_mutex_t file_write_mutex = PTHREAD_MUTEX_INITIALIZER;
+pthread_mutex_t thread_list_mutex = PTHREAD_MUTEX_INITIALIZER;
+
+// Signal handler to catch SIGINT and SIGTERM for graceful shutdown.
 void signal_handler(int signal) {
     if (signal == SIGINT || signal == SIGTERM) {
-        // Close the socket
         close(sockfd);
-        // Delete the file
         remove("/var/tmp/aesdsocketdata");
-        // Log the message
         syslog(LOG_INFO, "Caught signal, exiting");
-        // Close syslog
         closelog();
         exit(0);
+    } else if (signal == SIGALRM) {
+        char timestamp_str[100];
+        time_t current_time = time(NULL);
+        struct tm *local_time = localtime(&current_time);
+        strftime(timestamp_str, sizeof(timestamp_str), "timestamp:%a, %d %b %Y %H:%M:%S %z\n", local_time);
+        
+        pthread_mutex_lock(&mutex);
+        FILE *fp = fopen("/var/tmp/aesdsocketdata", "a");
+        if (fp == NULL) {
+            perror("fopen: Failed opening file.");
+        } else {
+            fprintf(fp, "%s", timestamp_str);
+            fclose(fp);
+        }
+        pthread_mutex_unlock(&mutex);
     }
 }
 
+void *connection_handler(void *socket_desc) {
+    int connfd = *(int *)socket_desc;
+    char *buffer = calloc(BUFFER_SIZE, sizeof(char));
+    
+    while (recv(connfd, buffer, BUFFER_SIZE, 0) > 0) {
+        pthread_mutex_lock(&mutex);
+        FILE *fp = fopen("/var/tmp/aesdsocketdata", "a");
+        if (fp == NULL) {
+            perror("fopen: Failed opening file.");
+            pthread_mutex_unlock(&mutex);
+            free(buffer);
+            close(connfd);
+            return NULL;
+        } else {
+            fprintf(fp, "%s", buffer);
+            fclose(fp);
+        }
+        pthread_mutex_unlock(&mutex);
+        
+        if (strchr(buffer, '\n') != NULL) {
+            FILE *fp_out = fopen("/var/tmp/aesdsocketdata", "r");
+            if (fp_out == NULL) {
+                perror("fpopen: Failed opening file.");
+                free(buffer);
+                close(connfd);
+                return NULL;
+            } else {
+                uint32_t bytes_read = 0;
+                while ((bytes_read = fread(buffer, 1, BUFFER_SIZE, fp_out)) > 0) {
+                    send(connfd, buffer, bytes_read, 0);
+                }
+                fclose(fp_out);
+            }
+        }
+        memset(buffer, 0, BUFFER_SIZE);
+    }
+    free(buffer);
+    close(connfd);
+    return NULL;
+}
+
+void add_thread(pthread_t tid) {
+    pthread_mutex_lock(&thread_list_mutex);
+
+    node_t *new_node = (node_t *)malloc(sizeof(node_t));
+    new_node->tid = tid;
+    new_node->next = head;
+    head = new_node;
+
+    pthread_mutex_unlock(&thread_list_mutex);
+}
+
+void cleanup_threads() {
+    pthread_mutex_lock(&thread_list_mutex);
+
+    node_t *current = head;
```

Comment:
@tiba6275 think about what `thread_list_mutex` is locking and it's reason for existing.  It may be related to `cleanup_threads` not being called.  If it's possible to modify the thread list in two threads you probably need locks whenever you access it, but if not you won't need locking.